### ISSUE TITLE
Do you have plan on supporting TCP+TLS to connect to backends?

### ISSUE THREAD

**Issue by didip** [yyyar/gobetween#38]
Hi all,
I have one question if you don't mind,
I see that the API supports TLS connection, but what about the TCP reverse proxy? Will you support TLS in the future?

**Comment by yyyar** [comment]
Hey @didip,
 We had thoughts to add TLS to proxy, but we postponed it because nobody really requested it.
 I think we can add it relatively quickly. Could you please describe your desired use case? 
Thanks!

**Comment by didip** [comment]
Oh man, thank you for the quick reply and interests.

My need is actually simple because you guys already handle all of the other features.

I'd love to be able to forward unsecured TCP payload to `gobetween` and it would handle the TLS handshake, round-robin balancing, and health checking to TCP+TLS backends.

`gobetween` pretty much handles everything already excepts TLS termination.

Also, It would be cool if you guys have settings for TLS caching and timeout just like Nginx. E.g.

```
ssl_session_cache   shared:SSL:20m;
ssl_session_timeout 4h;
```

I could use Nginx to solve my needs, but where's the fun in that :D

**Comment by illarion** [comment]
It looks like you mismatch 2 different approaches:
1) TLS termination (when clients negotiate with gobetween over TLS, but backends are plain tcp)
2) some other pattern (not named yet) - client is plain tcp, but backends are TLS.

The second looks like antipattern - it will consume a lot of computation power doing redundant work. So, is that what is really needed?

**Comment by didip** [comment]
Think about the case of end to end encryption with 3 nodes: `A -> B -> C -> db`

A: is edge node terminating TLS and has B as backend.

B: is intermediate node, also terminating TLS, that means A is also a TLS client talking to B.

C: is the node that talks to db, also terminating TLS, that means B is also a TLS client talking to C

It's true that this setup burns a lot of CPU
(which is why TLS client caching is important[1](https://golang.org/pkg/crypto/tls/#NewLRUClientSessionCache)),
but some companies care about this :)

**Comment by illarion** [comment]
@didip thank you for explanation. It seems that we need to implement TLS support for both ends of the proxy - talking to backends and talking to clients.

**Comment by didip** [comment]
Awesome! would you like me to close this question ticket? or leave it as is?

**Comment by yyyar** [comment]
@didip I've created 2 corresponding tickets: #39 and #40. Would they cover your case? 
I think #39 should be done first since it covers more popular use case I suppose.

**Comment by didip** [comment]
Yup! and yes, I agree termination should be handled first, testing is easier too:

1. Generate self-signed cert.

2. `echo "payload" | openssl s_client -connect localhost:3000 -tls1_2`

Also, ugh... I really don't want to be feature creep monster, but make sure to be able to choose TLS 1.0, 1.1, or strictly just 1.2. Example:

```
import "crypto/tls"

cfg := &tls.Config{
        MinVersion:               tls.VersionTLS12,
        CurvePreferences:         []tls.CurveID{tls.CurveP521, tls.CurveP384, tls.CurveP256},
        PreferServerCipherSuites: true,
        CipherSuites: []uint16{
            tls.TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,
            tls.TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA,
            tls.TLS_RSA_WITH_AES_256_GCM_SHA384,
            tls.TLS_RSA_WITH_AES_256_CBC_SHA,
        },
}
```

**Comment by yyyar** [comment]
Thanks, @didip 
Closing this ticket, let's use #39 and #40 for further discussions.

**Comment by illarion** [comment]
@didip please take a look at https://github.com/yyyar/gobetween/commit/18d0de2f405daffc51609f19fbb9d559edbb60d3

**Comment by didip** [comment]
Thanks for the reply! I will take a look at it.

**Comment by didip** [comment]
@illarion Looks really good, I could follow the logic from top to bottom, good job!

**Comment by illarion** [comment]
@didip https://github.com/yyyar/gobetween/wiki/Tls-Proxying

**Comment by didip** [comment]
This is awesome. Thank you for the hard work in considering this feature!

**Comment by yyyar** [comment]
@didip, you are welome! Let us know how it works for you.
it's now in master branch, we'll release it as a part of 0.4.0 after a while.
